### PR TITLE
Splice validation prose

### PR DESCRIPTION
**Pull request by f52985**
This PR splices the validation prose into the official document, mainly focusing on the validation of instructions.

The document contains the automatically generated and the manually written prose side by side, for the ease of comparison. Some of the notable differences are also noted inline.

@rossberg, please take a look and let us know if you have any idea to make it better.

### CODE REVIEW

**Comment by rossberg** [comment]
Sorry for the delay. I made a reading pass, and this looks fabulous! Here are my notes.

Some things we already talked about or you marked them in TODOs:

* Phrases like “t is contained in [numtype; vectype]” (e.g. select) are semantically off, because numtype and vectype are meta-variables that stand for many different types. Note sure there is a decent way to express it in set notation, so some prose “t is either a number type or a vector type” probably works better.

* Can we decompose more complex expressions (e.g. “unpack(zt)*” in struct.new) into auxiliary definitions? Perhaps approximating something akin to A-normal form wrt non-trivial constructs.

* Improve variable names (e.g. “valtype_u0” in select).


In addition, I have the following suggestions and questions:

* Can we unindent all bullet lists by one level, i.e., make the outermost (singular) level a non-bullet?

* Sentences should not start with a formula, since those may be lowercase letters (e.g. ref.test). Can we fix this by always inserting the type description of the expression, as in “The reference type rt matches rt'.”?

* “Matches” conditions are sometimes rendered with and sometimes without the type description, sometimes even in the same rule, which looks a bit odd (e.g. br_on_cast).

* Sometimes the meaning of “is” for equality reads a bit odd or unclear (could e.g. be interpreted as a Let). Maybe “equals” would be clearer?

* In various places (e.g. struct.new) “C.something[x] is defined _in the context_”, which is a bit redundant, as the C already implies that this is the context. Can we turn that into simply “C.something[x] is defined” (or simply “exists”, since that reads less funny when talking about defined types).

  In addition, there are many cases where the whole condition appears to be redundant, because it is implied by a following condition like “C.something[x] is y”. I noticed that this leads to a lot of bloat in the prose of many rules.

* When “binding” new variables in a comparison (e.g. “init” in local.set or “lim” in table.*), would it be possible to call that out that they are existential by saying something along the lines of “for some limit lim”?

* Instead of “x? is eps”, can we render that as “x is absent” for options (e.g. struct.get)?

* Instead of “default_t is val” can we write “default_t exists” when val isn’t actually used (e.g. struct/array.new_default)?

* In the rules for block instructions, the context composition expression has changed in the prose and does not fit what’s explained in the Note.

* Can we split the prose for catch clauses into separate rules?

* In the rule for ref.func, why is there the condition that C.refs[0] needs to be defined?

**Comment by presenthee** [comment]
Sorry for being late! I made some improvements based on your comments, and here are some remarks:

* Improve variable names (e.g. “valtype_u0” in select).

    &rarr; Unified variables are now indexed by their types, and they utilize shorter variable names (for example, "gt" for global types). 

    &rarr; I have also added a 'rename' option, which detaches the 'u' prefix from all unified variables (such as 'u0', 'u1', etc.) when enabled. This feature is considered experimental, as it can be useful for identifying unified variables in certain cases.
* When “binding” new variables in a comparison (e.g. “init” in local.set or “lim” in table.*), would it be possible to call that out that they are existential by saying something along the lines of “for some limit lim”?

    &rarr; As we discussed in the SpecTec meeting on 11/07, We’ll see if we can use animation process for this. This would be TODO after this PR is merged.
* Instead of “default_t is val” can we write “default_t exists” when val isn’t actually used (e.g. struct/array.new_default)?

    &rarr; I’m afraid that modifying the prose backend for this make the backend too specific for this purpose. Therefore, I added a relation ‘Defaultable’ with the prose hint. If you have any suggestions for this, please let me know.
* In addition, there are many cases where the whole condition appears to be redundant, because it is implied by a following condition like “C.something[x] is y”. I noticed that this leads to a lot of bloat in the prose of many rules.

    &rarr; Yes, Indeed. It seems to be partly due to sidecondtions. Currently validation proses are generated as almost one-to-one passion from premises to prose statements. 

    * For example, enabling sidecondtions creates conditions like 'if (|`t_1*`| = |`t_2*`|)' in the Resulttype_sub rule. 
    ```;; spec/wasm-3.0/3-typing.watsup:411.1-411.112
    relation Resulttype_sub: `%|-%<:%`(context, valtype*, valtype*)
      ;; spec/wasm-3.0/3-typing.watsup:414.1-416.37
      rule _{C : context, `t_1*` : valtype*, `t_2*` : valtype*}:
      `%|-%<:%`(C, t_1*{t_1 <- `t_1*`}, t_2*{t_2 <- `t_2*`})
      -- if (|`t_1*`| = |`t_2*`|)
      -- (Valtype_sub: `%|-%<:%`(C, t_1, t_2))*{t_1 <- `t_1*`, t_2 <- `t_2*`}
    ```
    * But when rendered, It seems to be quite redundant. We may need a way to ignore it.     
    
  

By the way, `oktypeidx` is not correctly rendered in the prose as below. Currently, AL expression `CaseE (OK, [VarE (x)])` is converted into EL FuseE expression `OK#($(x))!` in the backend-latex. However, it appears that the parentheses are removed in the `render_exp` function. Is there any workaround for this issue?

```OCaml
(*render_exp in render.ml of backend-latex*)
  | FuseE (e1, e2) ->
    (* TODO(2, rossberg): HACK for printing t.LOADn_sx (replace with invisible parens) *)
    let e2' = as_paren_exp (fuse_exp e2 true) in
    let es = e2' :: flatten_fuse_exp_rev e1 in
    String.concat "" (List.map (fun e -> "{" ^ render_exp env e ^ "}") (List.rev es))
```




Please take a look, and let me know if you have any further suggestions. :) @rossberg 

FYI, @f52985

**Review by rossberg — APPROVED** [review]
Excellent, looks good, thanks! LGTM modulo comments below.

> `OK#($(x))!` [...] Is there any workaround for this issue?

Ah thanks for noticing. Yes, the fuse operator removes one layer of parens, which is needed in some cases where the SpecTec grammar requires parens that we don't want to be rendered.

The simple way to force having parens anyway is by putting double parens in the show hint. And yes, oktypeidx (andoktypeidxnat) should be fixed that way.

File: spectec/spec/wasm-3.0/3-typing.watsup
Comment (presenthee) [reply spectec/spec/wasm-3.0/3-typing.watsup]: I guess that it's tricky to implement with relation hints.? so I referred to the official spec and added a definition for defaultable types. Now It looks like ` |- valtype defaultable` in the rule.

Comment (rossberg) [reply spectec/spec/wasm-3.0/3-typing.watsup]: Ah, right, there were no (regular) show hints for relations. I changed that, renaming the existing hints to `name` hints, and added the desired show hint to the above.

I also adapted a couple more uses of default side conditions. That required introducing the inverse Nondefaultable relation as well. It's unfortunate that we need those, but I have no better idea right now.

See https://github.com/Wasm-DSL/spectec/commit/9f341fb6182544c57bab1d596b5ec306cba166b9

Btw, there is one problem with the prose hints in that they hardcode Sphinx output (for cross references). Eventually, I'd like the prose backend to be able to produce prose with Latex markup as well (for wise users that don't want Sphinx), and then such hints would be broken. That's closely related to the problem of doing cross-references in prose. Not urgent, though.

Comment (presenthee) [reply spectec/spec/wasm-3.0/3-typing.watsup]: Thank you! Yes, clearly the prose hints need improvement, as they are implemented naively. We're in the process of listing the work required for cross-references, and I'll add this as well.

**Pull request merged**
